#1.1 

In [ ]:
#1.1 
import numpy as np

aux = [] #storage

def DivMed(qt,a):
  xi = 1 # start in 1
  for k in range(qt):
    x = (xi + a/xi)/2 #function
    k+=1
    aux.append(x)
  print(aux)

#erro relativo se dá por: 
#|x0 - x1| / |x0| 
def calculateRelativeError(aux):
  listErrorRelative = []
  for j in aux:
    relativeError = np.trunc(j) #aproximando o valor
    aprox =  (( j - relativeError)/np.abs(j))
    listErrorRelative.append(aprox)
  print(listErrorRelative)

#1.2

In [ ]:
#1.2
e = 1
while (1+e<=1):
  if (1+ e <= 1):
    e = 2 * e
  else:
    e = e/2
  print(e)


In [ ]:
# Com laço de repetição
e = 1
for k in range(100):
  if (1+ e <= 1):
    e = 2 * e
  else:
    e = e/2
  print(e)


0.5
0.25
0.125
0.0625
0.03125
0.015625
0.0078125
0.00390625
0.001953125
0.0009765625
0.00048828125
0.000244140625
0.0001220703125
6.103515625e-05
3.0517578125e-05
1.52587890625e-05
7.62939453125e-06
3.814697265625e-06
1.9073486328125e-06
9.5367431640625e-07
4.76837158203125e-07
2.384185791015625e-07
1.1920928955078125e-07
5.960464477539063e-08
2.9802322387695312e-08
1.4901161193847656e-08
7.450580596923828e-09
3.725290298461914e-09
1.862645149230957e-09
9.313225746154785e-10
4.656612873077393e-10
2.3283064365386963e-10
1.1641532182693481e-10
5.820766091346741e-11
2.9103830456733704e-11
1.4551915228366852e-11
7.275957614183426e-12
3.637978807091713e-12
1.8189894035458565e-12
9.094947017729282e-13
4.547473508864641e-13
2.2737367544323206e-13
1.1368683772161603e-13
5.684341886080802e-14
2.842170943040401e-14
1.4210854715202004e-14
7.105427357601002e-15
3.552713678800501e-15
1.7763568394002505e-15
8.881784197001252e-16
4.440892098500626e-16
2.220446049250313e-16
1.1102230246251565e-16
2.22

In [ ]:
#teste
def machineEpsilon(func=float):
    machine_epsilon = func(1)
    while func(1)+func(machine_epsilon) != func(1):
        machine_epsilon_last = machine_epsilon
        machine_epsilon = func(machine_epsilon) / func(2)
    return machine_epsilon_last

In [ ]:
machineEpsilon(np.float16)

0.000977

In [ ]:
#Usando o numpy para pegar o epsilon da máquina
np.finfo(float(1))

finfo(resolution=1e-15, min=-1.7976931348623157e+308, max=1.7976931348623157e+308, dtype=float64)

#1.3

In [ ]:
#1.3
def converge(n):
  lista = [1/3]
  for k in range(n):
    j = (4*lista[-1]) - 1
    lista.append(j)
  print(lista)

In [ ]:
converge(5)

[0.3333333333333333, 0.33333333333333326, 0.33333333333333304, 0.33333333333333215, 0.3333333333333286, 0.3333333333333144]


# A série iterativa sempre irá divergir com variações muito pequenas.
Para n = 5 termos: [0.3333333333333333, 0.33333333333333326, 0.33333333333333304, 0.33333333333333215, 0.3333333333333286, 0.3333333333333144].

#1.4

In [ ]:
#1.4
def expi(u):
  if u > 0:
    x = (1/u)
    j = np.exp(x)
    k = (j / (1 + j))
    print(k)

In [ ]:
expi(0.001)

nan


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:5: RuntimeWarning: overflow encountered in exp
  """
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:6: RuntimeWarning: invalid value encountered in double_scalars
  


In [ ]:
def expa(w):
  if w > 0:
    y = (-1/w)
    g = np.exp(y)
    o = (1/ (g+1))
    print(o)

In [ ]:
expa(0.001)

1.0


A valor da expressão 1 dá um overflow, por isso não conseguimos utilizar valores pequenos na primeira expressão.

#1.5

In [ ]:
p1 = 10**(-12)
p2 = 10**(-15)
p3 = 10**(-17)

def expi(x):
  exps = ((1+x) - 1) / x
  return(exps)

In [ ]:
np.format_float_scientific(expi(p1))

'1.000088900582341e+00'

In [ ]:
np.format_float_scientific(expi(p2))

'1.1102230246251565e+00'

In [ ]:
np.format_float_scientific(expi(p3))

'0.e+00'

x = 10**−17 ( no código o p3 = 10**(-17) , gera um underflow. Quando tentamos representar um valor muito pequeno esse erro acontece.)